### ISSUE TITLE
New Keyboard Shortcuts

### ISSUE THREAD

**Issue by ndom91** [gitbutlerapp/gitbutler#4734]
We're looking for input on some more keyboard shortcuts. We want to align with other similar apps and implement keyboard shortcuts that will be familiar to users. We're currently considering adding the following:

- [ ] `ctrl + a` to select all files in filelist in commit mode
- [ ] `space` to toggle select/unselect on focused file
- [ ] `esc` to close an opened file diff preview
- [ ] `ctrl + ?` to open a keyboard shortcut overview menu

Also potentially `ctrl + z/y` to travel back and forth in the oplog aka undo stack. 

Please add any other ideas below :grin:

**Comment by freak4pc** [comment]
Just wanted to report that using up/down keys to navigate between files doesn't work - not sure if it's planned to not work or it's a bug, but that's the behavior in other Git UI tools :)

**Comment by ndom91** [comment]
Thanks for chiming in! Yeah those sound sensible as well 👌

**Comment by freak4pc** [comment]
> > Just wanted to report that using the up/down keys to navigate between files doesn't work. Not sure if it's intentional or a bug, but that's how it works in other Git UI tools. :)
> 
> @freak4pc This is indeed a bug. The functionality should be working, so I'm wondering why it isn't on your end. If possible, could you please create a separate issue for this? You can do so using [this link](https://github.com/gitbutlerapp/gitbutler/issues/new?assignees=&labels=bug&projects=&template=1_bug.yml).
> 
>  Screen.Recording.2024-08-26.at.15.22.52.mov

Thanks, I actually meant when being focused on a specific commit, not on the staged workspace changes

**Comment by PavelLaptev** [comment]
> Thanks, I actually meant when being focused on a specific commit, not on the staged workspace changes

@freak4pc, thank you for clarifying this. Yes, it doesn’t work at the moment. It's something we need to fix.

**Comment by H1ghBre4k3r** [comment]
I am currently thinking if hitting `esc` while in the settings (both, project and app settings) should bring you back to the previous screen from where you opened the settings (similar to the small arrow on the upper left hand side).

**Comment by ndom91** [comment]
Note for ourselves for the future: Navigating `up`/`down` with the keyboard arrows in the Filelist sometimes behaves very strangely. For example, skipping items, stopping at "incorrect" items, etc.

**Comment by PavelLaptev** [comment]
> Note for ourselves for the future: Navigating `up`/`down` with the keyboard arrows in the Filelist sometimes behaves very strangely. For example, skipping items, stopping at "incorrect" items, etc.

Can you please also attach a video capture or describe how to reproduce this.

**Comment by ndom91** [comment]
There's not much to record, for example in this scenario:

> ![image](https://github.com/user-attachments/assets/bd31a22a-f4f8-4fb8-b320-aac364f40add)

If I select the top file, `controller.js`, I can only navigate down to `index.js` with the keyboard down arrow. Then it gets stuck. 

Similarly, if I click on the bottom item, `test.txt`, and press up on the keyboard arrow, it jumps to `index.js` (skipping `package.json`) and then gets stuck on `index.js`.

**Comment by esolitos** [comment]
A few notes:

* `CMD + \` (toggle sidebar) is already a default shortcut for both 1Password as well as [Launchbar](https://www.obdev.at/products/launchbar/index.html). 😓 Would it be possible to change this?
* There is no shortcut (or is not shown) to select the "project switcher". I would argue that `CMD + T` is a common one, but in GB is already assigned to a _very useful_ (sarcasm :D) theme switching. 😅 Sarcasm aside, ideally in this context there should be those shortcuts:
    * One to open the project switcher
    * One that switches to the "previous project", a typical usage of this could be _"making a fix on a dependent project just to switch back to the main one after"_

**Comment by aaronzhongg** [comment]
Came here looking for a project switcher shortcut, naturally went for `CMD + P`, another nice quality of life one would be to pull upstream commits

**Comment by sminnee** [comment]
Do you plan on spending some time on keyboard shortcuts in the near future? It's a great app but this really holds it back—right now, at least with the new UI, it doesn't feel like keyboard shortcuts have been given any attention, other than they greyed out keyboard shortcuts icon.

**Comment by PavelLaptev** [comment]
@sminnee Thanks for the feedback! We actually have a [large keyboard shortcut map](https://docs.google.com/spreadsheets/d/1Ed88h2WYg2OGiJjKqnDChrB0Wr1auLwg8RAm5jm_1pw) planned, but as you mentioned, the new UI is still in development, so we held off on the implementation. I’ll try to push this up and see what we can already start adding.
If there are specific shortcuts you’d like to see, feel free to share them!

**Comment by sminnee** [comment]
The main area that i would like to see keyboard shortcuts is when staging files for a commit. 

The critical:
 * Being able to use up/down arrows to cycle between files (this _may_ be working but maybe it's recent or buggy?)
 * Space or enter to toggle wether the file is included

The nice to have:
 * Being able to tab to the chunks of a single file and then use a similar interaction to stage/unstage chunks
 * It would also be great to go one level further than then do the same with single lines, but that would probably be the least essential.

Allowing up/down arrow navigation of other lists would be great. Right now when I select one and press up/down i just see a blue horizontal line a bit like it's trying to select something
 * The individual commits in a branch
 * The individual files in a commit

Being able to override the shortcuts with my own bindings is pretty low priority. Just knowing what the defaults are for me to learn would be okay.


----

TL;DR: "Move selection up", "'Move selection down", and "Toggle selection" on your keyboard shortcut map (rows 94, 95, 107)

**Comment by tomgilder** [comment]
I would love the ability to commit _and_ push at the same time from here:

<img width="349" height="219" alt="Image" src="https://github.com/user-attachments/assets/3c611b0e-103d-4981-be47-83d1fb3675f5" />

Maybe option+command+return?

**Comment by dev-ankit** [comment]
May I just suggest a key bindings config like we see in vim, Ghostty and others. basically a TOML file.
In the beginning you can even leave most of these empty and the users can assign these themselves and as you think about good defaults you can add them

Getting a way to let people assign this would be really helpful

**Comment by Byron** [comment]
I fully agree, in the end, everything should be customisable.

**Comment by elv1n** [comment]
The tool seems to be working exactly like I need for agentic workflows, navigation and UI is there but without shortcuts and its customizations is a no go.

**Comment by Byron** [comment]
@elv1n What's easily missed in the current GB graphical UI is that one can press `f` to select any clickable item visible on screen.
Meantime, the TUI by nature is fully controllable by keyboard, and the upcoming electron based GUI will have keyboard navigation natively from the start. It really is getting there.
